### PR TITLE
Refactoring of Cron-Scripts

### PR DESCRIPTION
**Pull request by kaustabhbarman**

### CODE REVIEW

File: tests/library/Application/Job/CleanTemporariesJobTest.php [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+ */
+
+require_once dirname(__FILE__) . '/../../../scripts/cron/OpusDocumentMock.php';
+require_once dirname(__FILE__) . '/../../../../library/Application/Job/CleanTemporariesJob.php';
+
+use Opus\Date;
+use Opus\Document;
+use Opus\Model\NotFoundException;
+
+class CleanTemporaries extends ControllerTestCase
```
Comment (j3nsch) [inline tests/library/Application/Job/CleanTemporariesJobTest.php:41]: The test class needs to have the same name as the file, so it should be `CleanTemporatiesJobTest`

File: tests/library/Application/Job/CleanTemporariesJobTest.php [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ * @category    Application Unit Test
+ * @package     Application
+ * @author      Kaustabh Barman <barman@zib.de>
+ * @copyright   Copyright (c) 2021, OPUS 4 development team
+ * @license     http://www.gnu.org/licenses/gpl.html General Public License
+ */
+
+require_once dirname(__FILE__) . '/../../../scripts/cron/OpusDocumentMock.php';
```
Comment (j3nsch) [inline tests/library/Application/Job/CleanTemporariesJobTest.php:34]: We don't use `require_once` in our regular unit tests. It should not be necessary.

Comment (j3nsch) [reply tests/library/Application/Job/CleanTemporariesJobTest.php:34]: If you need to move the class `OpusDocumentMock`, do it. We already have folders for classes supporting tests.

File: tests/library/Application/Job/CleanTemporariesJobTest.php [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+require_once dirname(__FILE__) . '/../../../../library/Application/Job/CleanTemporariesJob.php';
```
Comment (j3nsch) [inline tests/library/Application/Job/CleanTemporariesJobTest.php:35]: Definitely not. Please take a look at the other unit tests.

File: tests/library/Application/Job/CleanTemporariesJobTest.php [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        $this->doc->store();
+    }
+
+    public function testRun()
+    {
+        $this->changeDocumentDateModified($this->doc, 3);
+        $this->job->run();
+
+        $doc = Document::get($this->doc->getId());
+        $this->setExpectedException(NotFoundException::class);
```
Comment (j3nsch) [inline tests/library/Application/Job/CleanTemporariesJobTest.php:61]: You need to declare the expected exception before performing the operation that throws it, so switch lines 60 and 61.

File: scripts/cron/cron-send-notification.php [hunk 5]
```diff
@@ -38,21 +38,9 @@
...
-$jobrunner->setLimit(100);
-
-$mailWorker = new MailNotification(null, false);
-$mailWorker->setLogger(Log::get());
-
-$jobrunner->registerWorker($mailWorker);
-
-$jobrunner->run();
+//$job = new \library\Application\Job\SendNotificationJob();
+$job = new SendNotificationJob();
```
Comment (j3nsch) [inline scripts/cron/cron-send-notification.php:45]: The name of the class is `Application_Job_SendNotificationJob`.

File: library/Application/Job/SendNotificationJob.php [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+require_once('JobInterface.php');
+
+use Opus\Job\Runner;
+use Opus\Job\Worker\MailNotification;
+use Opus\Log;
+
+/**
+ * Class for sending notification
+ */
+class SendNotificationJob implements JobInterface
```
Comment (j3nsch) [inline library/Application/Job/SendNotificationJob.php:42]: The name of the class is `Application_Job_SendNotificationJob`.

File: library/Application/Job/CheckConsistensyJob.php [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+require_once('JobInterface.php');
+
+use Opus\Job\Runner;
+use Opus\Log;
+use Opus\Search\Task\ConsistencyCheck;
+
+/**
+ * Class to check consistency
+ */
+class Application_Job_CheckConsistensyJob implements JobInterface
```
Comment (j3nsch) [inline library/Application/Job/CheckConsistensyJob.php:42]: "Consistency" with a "c" before the "y"

**Review by j3nsch — CHANGES_REQUESTED** [review]
I stop the review without looking at the classes in more detail. At the moment code was just moved. That is fine for now. The classes should have tests, but for some classes, the once that use `Runner` we need to think about what the purpose of each classes is and if the current design makes sense. Are we wrapping a job in a job here. Is it necessary? Why? What next steps should be taken?

File: Jenkinsfile [hunk 8]
```diff
@@ -1,3 +1,5 @@
+
```
[the remark this answers is not in the document]
Comment (kaustabhbarman) [reply Jenkinsfile:1]: It was not intentional. Must have been by mistake.

**Comment by j3nsch** [comment]
Modified title of PR so it can be understood without effort.

**Comment by kaustabhbarman** [comment]
> I stop the review without looking at the classes in more detail. At the moment code was just moved. That is fine for now. The classes should have tests, but for some classes, the once that use `Runner` we need to think about what the purpose of each classes is and if the current design makes sense. Are we wrapping a job in a job here. Is it necessary? Why? What next steps should be taken?

In the classes that uses `Runner`, a worker class is being registered first in `Runner` class using `registerWorker()` function. The `run` function of `Runner` fetches jobs, creates new jobs, execute them by calling `work()` function of the worker class and lastly, removes it from the jobs table on success. 
This operation consists of few steps like setting the logger, delay, limit and registering the worker and currently, theses steps are being wrapped in the job class so that they can all be done with a single call. 

In my opinion, as long as we don't want to change some of the arguments like a different delay or limit, this design makes sense. But if we are to make these arguments dynamic, we might as well call each function like `setDelay()` or `setLimit()` instead of wrapping them in another job wrapper and passing these arguments as class parameters.

File: library/Application/Job/CheckWorkspaceFilesJob.php [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        }
+
+        echo "INFO: Checked a total of $count entries with ".round($count / (microtime(true) - $this->startTime))." entries/seconds.\n";
+
+        if ($errors === 0) {
+            exit(0);
+        }
+
+        echo "ERROR: Found $errors ERRORs in workspace files directory '$filesPath'!\n";
+        exit(1);
```
Comment (j3nsch) [inline library/Application/Job/CheckWorkspaceFilesJob.php:85]: Exit statements would be bad if we want to execute this class in some other context. There should not be any exit statements in the job classes.

**Pull request merged**